libdnf/goal/Goal.cpp

```diff
@@ -1141,6 +1141,10 @@ Goal::Impl::allowUninstallAllButProtected(Queue *job, DnfGoalActions flags)
     if (kernel > 0)
         protectedPkgs->set(kernel);
 
+    if (DNF_NO_PROTECTED & flags) {
+        protectedPkgs.reset(new PackageSet(sack));
+    }
```

Comment:
DNF_NO_PROTECTED flag is passed only into goal run. I would suggest that it should have only temporal effect on goal.run() and not modify global values in Goal. See DNF_ALLOW_UNINSTALL flag.
I would recommend to remove the `if` and replacing it by conditional application of protected packages.
We can discuss it in more details personally.